### PR TITLE
Don't write 0 to the end of strings in embind

### PR DESCRIPTION
**Pull request by jgravelle-google**
This is unsafe because it's one past the end of the string, and in threaded contexts it can cause data races.

### CODE REVIEW

File: src/runtime_strings_extra.js [hunk 1]
```diff
@@ -39,7 +39,8 @@ var UTF16Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf-16l
 #endif // TEXTDECODER
 #endif // TEXTDECODER == 2
 
-function UTF16ToString(ptr) {
+function UTF16ToString(ptr, maxCharsToRead) {
```
Comment (kripken) [inline src/runtime_strings_extra.js:42]: The utf8 version calls the param `maxBytesToRead`, not chars, is the difference intentional?

File: src/runtime_strings_extra.js [hunk 2]
```diff
@@ -39,7 +39,8 @@ var UTF16Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf-16l
...
+  if (maxCharsToRead === undefined) maxCharsToRead = Infinity;
```
Comment (kripken) [inline src/runtime_strings_extra.js:43]: I think we should be consistent with how `UTF8ArrayToString` does things, it has that trick,
```javascript
  // (As a tiny code save trick, compare endPtr against endIdx using a negation, so that undefined means Infinity)
```
We should either do them all that way, or none, I think, so it's consistent one way or the other. I lean towards doing the trick, with that documentation everywhere.

Comment (jgravelle-google) [reply src/runtime_strings_extra.js:43]: Makes sense (and I'd missed that comment so I was more befuddled than I needed to be when I first saw it), consistency makes sense.

File: src/runtime_strings_extra.js (hunk 1 above)
Comment (jgravelle-google) [reply src/runtime_strings_extra.js:42]: Yes, though I'm not sure what's best here. I have no idea how people expect to use UTF16 or 32 strings in practice. The C++ wstrings have a length that seems accurate? My understanding of unicode encodings is honestly pretty shaky.

This is also why I didn't change documentation; I really don't want people to use these. Which *maybe* isn't the best approach, granted. Does it make sense to leave a comment here saying "hey don't use this" but still keep the parameter to fix the issue? Is there a better way to avoid having to write a 0 byte?

The other question is a consistency one: does it make sense to change utf8 to maxChars, or is it actually bytes? I matched the names to my understanding, so thank you for noticing that.

**Comment by jgravelle-google** [comment]
> How well is this tested?

There's some embind tests in test_core, and other.test_embind is pretty thorough. other.test_embind fails for all of the string types if I just remove the 0-byte-write, and passes with the rest of the change. That seemed reasonable to me, but there are probably more corner-cases in utf encodings that we don't test.

File: src/runtime_strings_extra.js (hunk 1 above)
Comment (kripken) [reply src/runtime_strings_extra.js:42]: Good points - I'm not sure about those either.

We could take the time to investigate this in depth, but I wonder if that's worth it? The by far important use case is UTF8. I'm not sure if 16 and 32 are even that important. If we do a little searching and don't find a reason not to, just saying embind only does UTF8 for now might be an option.

**Comment by kripken** [comment]
I think we should add proper embind + pthreads testing eventually, and for all the string types, but that doesn't need to block this PR. But the other comments above should be addressed I think.

File: src/runtime_strings_extra.js (hunk 1 above)
Comment (dschuff) [reply src/runtime_strings_extra.js:42]: It looks like similar Dart and .NET APIs use bytes, so maybe we could do that too.

**Comment by dschuff** [comment]
Yeah, test_embind looks pretty reasonable overall. Maybe we could address the 2 comments in code but not update the documents (i.e. leave the extra 'bytes' option undocumented for now) until we add a test that exercises it in a future PR?

File: src/runtime_strings_extra.js [hunk 3]
```diff
@@ -117,16 +119,17 @@ function lengthBytesUTF16(str) {
...
   assert(ptr % 4 == 0, 'Pointer passed to UTF32ToString must be aligned to four bytes!');
 #endif
   var i = 0;
+  if (maxCharsToRead === undefined) maxCharsToRead = Infinity;
 
   var str = '';
   while (1) {
     var utf32 = {{{ makeGetValue('ptr', 'i*4', 'i32') }}};
-    if (utf32 == 0) return str;
+    if (utf32 == 0 || i >= maxCharsToRead) return str;
```
Comment (juj) [inline src/runtime_strings_extra.js:132]: Instead of having a `while(1)` with `i >= maxCharsToRead`, it would be better to move the index check to the `while`.

These changes to UTF16ToString and UTF32ToString regress code size. I am still keeping my hopes up that Closure compiler would one day be able to optimize it (https://github.com/google/closure-compiler/issues/3193, https://github.com/google/closure-compiler/issues/3201 and https://github.com/google/closure-compiler/issues/3223). Could you change this to use the same compare against NaN/undefined trick that UTF8ToString uses to save code size and make the code (perhaps) optimizable by Closure in the future?

File: src/runtime_strings_extra.js [hunk 4]
```diff
@@ -64,7 +67,7 @@ function UTF16ToString(ptr) {
     var str = '';
     while (1) {
       var codeUnit = {{{ makeGetValue('ptr', 'i*2', 'i16') }}};
-      if (codeUnit == 0) return str;
+      if (codeUnit == 0 || i == maxBytesToRead / 2) return str;
```
Comment (juj) [inline src/runtime_strings_extra.js:70]: Should this not be `i == maxIdx`? Having a division in the hot per-characterr loop would be bad for performance.

File: src/runtime_strings_extra.js [hunk 5]
```diff
@@ -117,16 +120,18 @@ function lengthBytesUTF16(str) {
...
 #if ASSERTIONS
   assert(ptr % 4 == 0, 'Pointer passed to UTF32ToString must be aligned to four bytes!');
 #endif
   var i = 0;
 
   var str = '';
-  while (1) {
+  // If maxBytesToRead is not passed explicitly, it will be undefined, and this
+  // will always evaluate to true. This saves on code size.
+  while (!(i >= maxBytesToRead / 4)) {
```
Comment (juj) [inline src/runtime_strings_extra.js:132]: Likewise here, having a per-loop iteration division is not good for perf.

File: src/runtime_strings_extra.js (hunk 4 above)
Comment (juj) [reply src/runtime_strings_extra.js:70]: Also note the inconsistency in implementation between UTF16ToString() and UTF32ToString: this retains a `while(1)` whereas UTF32ToString switched to a `while (!(i >= maxBytesToRead / 4)) {`. This variant is better for code size and less branches (termination test is preserved in one if, whereas it is in two parts in UTF32ToString), let's use UTF16ToString() format in both?

File: src/runtime_strings_extra.js [hunk 6]
```diff
@@ -117,16 +120,18 @@ function lengthBytesUTF16(str) {
...
     var utf32 = {{{ makeGetValue('ptr', 'i*4', 'i32') }}};
```
Comment (juj) [inline src/runtime_strings_extra.js:133]: These makeGetValues could be demoted to direct HEAPU32/HEAPU16 reads for better performance and smaller code size, although could leave that for a later day as well.

Comment (kripken) [reply src/runtime_strings_extra.js:133]: The compiler turns those `{{{ makeGetValue }}}` things into `HEAP32` at compile time, so there should be no speed or size difference.

File: src/runtime_strings_extra.js (hunk 5 above)
Comment (kripken) [reply src/runtime_strings_extra.js:132]: I think it is safe to assume the JS VM will do loop invariant code motion on such things, it's been standard for many years since the JS speed wars. And it can be more compact this way instead of defining a new variable (unless we also need that variable elsewhere).

File: src/runtime_strings_extra.js (hunk 4 above)
Comment (kripken) [reply src/runtime_strings_extra.js:70]: +1 for a followup to make the code more consistent.

File: src/runtime_strings_extra.js (hunk 6 above)
Comment (juj) [reply src/runtime_strings_extra.js:133]: I don't think it is able to optimize away the *4, so it will end up with a *4 in each iteration. Manually iterating over indices instead of iterating over bytes will be shorter code size wise and also faster too.

Comment (kripken) [reply src/runtime_strings_extra.js:133]: That's a good question, yeah, maybe not. Sounds good to iterate over `HEAP32` indexes in a followup.
